File: js/packages/openinference-core/src/trace/trace-config/maskingRules.ts
import { AttributeValue } from "@opentelemetry/api";
import { REDACTED_VALUE } from "./constants";
import { SemanticConventions } from "@arizeai/openinference-semantic-conventions";
import { MaskingRule, MaskingRuleArgs } from "./types";

/**
 * Masks (redacts) input text in LLM input messages.
 * Will mask information stored under the key `llm.input_messages.[i].message.content`.
 * @example
 * ```typescript
 *  maskInputTextRule.condition({
 *      config: {hideInputText: true},
 *      key: "llm.input_messages.[i].message.content"
 *  }) // returns true so the rule applies and the value will be redacted
 */
const maskInputTextRule: MaskingRule = {
  condition: ({ config, key }) =>
    config.hideInputText &&
    key.includes(SemanticConventions.LLM_INPUT_MESSAGES) &&
    key.includes(SemanticConventions.MESSAGE_CONTENT) &&
    !key.includes(SemanticConventions.MESSAGE_CONTENTS),
  action: () => REDACTED_VALUE,
};

/**
 * Masks (redacts) output text in LLM output messages.
 * Will mask information stored under the key `llm.output_messages.[i].message.content`.
 * @example
 * ```typescript
 *  maskOutputTextRule.condition({
 *      config: {hideOutputText: true},
 *      key: "llm.output_messages.[i].message.content"
 *  }) // returns true so the rule applies and the value will be redacted
 * ```
 */
const maskOutputTextRule: MaskingRule = {
  condition: ({ config, key }) =>
    config.hideOutputText &&
    key.includes(SemanticConventions.LLM_OUTPUT_MESSAGES) &&
    key.includes(SemanticConventions.MESSAGE_CONTENT) &&
    !key.includes(SemanticConventions.MESSAGE_CONTENTS),
  action: () => REDACTED_VALUE,
};

/**
 * Masks (redacts) input text content in LLM input messages.
 * Will mask information stored under the key `llm.input_messages.[i].message.contents.[j].message_content.text`.
 * @example
 * ```typescript
 *  maskOutputTextRule.condition({
 *      config: {hideInputText: true},
 *      key: "llm.input_messages.[i].message.contents.[j].message_content.text"
 *  }) // returns true so the rule applies and the value will be redacted
 */
const maskInputTextContentRule: MaskingRule = {
  condition: ({ config, key }) =>
    config.hideInputText &&
    key.includes(SemanticConventions.LLM_INPUT_MESSAGES) &&
    key.includes(SemanticConventions.MESSAGE_CONTENT_TEXT),
  action: () => REDACTED_VALUE,
};

/**
 * Masks (redacts) output text content in LLM output messages.
 * @example
 * ```typescript
 *  maskOutputTextRule.condition({
 *      config: {hideOutputText: true},
 *      key: "llm.output_messages.[i].message.contents.[j].message_content.text"
 *  }) // returns true so the rule applies and the value will be redacted
 */
const maskOutputTextContentRule: MaskingRule = {
  condition: ({ config, key }) =>
    config.hideOutputText &&
    key.includes(SemanticConventions.LLM_OUTPUT_MESSAGES) &&
    key.includes(SemanticConventions.MESSAGE_CONTENT_TEXT),
  action: () => REDACTED_VALUE,
};

/**
 * Masks (removes) input images in LLM input messages.
 * @example
 * ```typescript
 *  maskOutputTextRule.condition({
 *      config: {hideInputImages: true},
 *      key: "llm.input_messages.[i].message.contents.[j].message_content.image"
 *  }) // returns true so the rule applies and the value will be removed
 */
const maskInputImagesRule: MaskingRule = {
  condition: ({ config, key }) =>
    config.hideInputImages &&
    key.includes(SemanticConventions.LLM_INPUT_MESSAGES) &&
    key.includes(SemanticConventions.MESSAGE_CONTENT_IMAGE),
  action: () => undefined,
};

function isBase64Url(url?: AttributeValue): boolean {
  return (
    typeof url === "string" &&
    url.startsWith("data:image/") &&
    url.includes("base64")
  );
}

/**
 * Masks (redacts) base64 images that are too long.
 *  * @example
 * ```typescript
 *  maskOutputTextRule.condition({
 *      config: {base64ImageMaxLength: 10},
 *      key: "llm.input_messages.[i].message.contents.[j].message_content.image.url",
 *      value: "data:image/base64,verylongbase64string"
 *  }) // returns true so the rule applies and the value will be redacted
 */
const maskLongBase64ImageRule: MaskingRule = {
  condition: ({ config, key, value }) =>
    typeof value === "string" &&
    isBase64Url(value) &&
    value.length > config.base64ImageMaxLength &&
    key.includes(SemanticConventions.LLM_INPUT_MESSAGES) &&
    key.includes(SemanticConventions.MESSAGE_CONTENT_IMAGE) &&
    key.endsWith(SemanticConventions.IMAGE_URL),
  action: () => REDACTED_VALUE,
};

/**
 * Masks (removes) embedding vectors.
 *  * @example
 * ```typescript
 *  maskOutputTextRule.condition({
 *      config: {hideEmbeddingVectors: true},
 *      key: "embedding.embeddings.[i].embedding.vector"
 *  }) // returns true so the rule applies and the value will be redacted
 */
const maskEmbeddingVectorsRule: MaskingRule = {
  condition: ({ config, key }) =>
    config.hideEmbeddingVectors &&
    key.includes(SemanticConventions.EMBEDDING_EMBEDDINGS) &&
    key.includes(SemanticConventions.EMBEDDING_VECTOR),
  action: () => undefined,
};

/**
 * A list of {@link MaskingRule}s that are applied to span attributes to either redact or remove sensitive information.
 * The order of these rules is important as it can ensure appropriate masking of information
 * Rules should go from more specific to more general so that things like `llm.input_messages.[i].message.content` are masked with {@link REDACTED_VALUE} before the more generic masking of `llm.input_messages` might happen with `undefined` might happen.
 */
const maskingRules: MaskingRule[] = [
  {
    condition: ({ config, key }) =>
      config.hideInputs && key === SemanticConventions.INPUT_VALUE,
    action: () => REDACTED_VALUE,
  },
  {
    condition: ({ config, key }) =>
      config.hideInputs && key === SemanticConventions.INPUT_MIME_TYPE,
    action: () => undefined,
  },
  {
    condition: ({ config, key }) =>
      config.hideOutputs && key === SemanticConventions.OUTPUT_VALUE,
    action: () => REDACTED_VALUE,
  },
  {
    condition: ({ config, key }) =>
      config.hideOutputs && key === SemanticConventions.OUTPUT_MIME_TYPE,
    action: () => undefined,
  },
  {
    condition: ({ config, key }) =>
      (config.hideInputs || config.hideInputMessages) &&
      key.includes(SemanticConventions.LLM_INPUT_MESSAGES),
    action: () => undefined,
  },
  {
    condition: ({ config, key }) =>
      (config.hideOutputs || config.hideOutputMessages) &&
      key.includes(SemanticConventions.LLM_OUTPUT_MESSAGES),
    action: () => undefined,
  },
  maskInputTextRule,
  maskOutputTextRule,
  maskInputTextContentRule,
  maskOutputTextContentRule,
  maskInputImagesRule,
  maskLongBase64ImageRule,
  maskEmbeddingVectorsRule,
];

/**
 * A function that masks (redacts or removes) sensitive information from span attributes based on the trace config.
 * @param config The {@link TraceConfig} to use to determine if the value should be masked
 * @param key The key of the attribute to mask
 * @param value The value of the attribute to mask
 * @returns The redacted value or undefined if the value should be masked, otherwise the original value
 */
export function mask({
  config,
  key,
  value,
}: MaskingRuleArgs): AttributeValue | undefined {
  for (const rule of maskingRules) {
    if (rule.condition({ config, key, value })) {
      return rule.action();
    }
  }
  return value;
}

File: java/openinference-instrumentation/src/main/java/com/arize/instrumentation/trace/config/MaskingRules.java
package com.arize.instrumentation.trace.config;

import com.arize.instrumentation.TraceConfig;
import com.arize.semconv.trace.SemanticConventions;
import io.opentelemetry.api.common.Value;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import lombok.Builder;
import lombok.Getter;

/**
 * Arguments for masking rules
 */
@Getter
@Builder
class MaskingRuleArgs {
    private TraceConfig config;
    private String key;
    private Value value;
}

/**
 * A masking rule that defines when and how to mask attributes
 */
@Getter
class MaskingRule {
    private final Predicate<MaskingRuleArgs> condition;
    private final Function<MaskingRuleArgs, Value> action;

    public MaskingRule(Predicate<MaskingRuleArgs> condition, Function<MaskingRuleArgs, Value> action) {
        this.condition = condition;
        this.action = action;
    }
}

/**
 * Utility class for masking sensitive information in span attributes
 */
class MaskingUtils {

    private static final Value<String> REDACTED_VALUE = Value.of("[REDACTED]");

    /**
     * Masks (redacts) input text in LLM input messages.
     * Will mask information stored under the key `llm.input_messages.[i].message.content`.
     */
    private static final MaskingRule maskInputTextRule = new MaskingRule(
            args -> args.getConfig().isHideInputText()
                    && args.getKey().contains(SemanticConventions.LLM_INPUT_MESSAGES)
                    && args.getKey().contains(SemanticConventions.MESSAGE_CONTENT)
                    && !args.getKey().contains(SemanticConventions.MESSAGE_CONTENTS),
            args -> REDACTED_VALUE);

    /**
     * Masks (redacts) output text in LLM output messages.
     * Will mask information stored under the key `llm.output_messages.[i].message.content`.
     */
    private static final MaskingRule maskOutputTextRule = new MaskingRule(
            args -> args.getConfig().isHideOutputText()
                    && args.getKey().contains(SemanticConventions.LLM_OUTPUT_MESSAGES)
                    && args.getKey().contains(SemanticConventions.MESSAGE_CONTENT)
                    && !args.getKey().contains(SemanticConventions.MESSAGE_CONTENTS),
            args -> REDACTED_VALUE);

    /**
     * Masks (redacts) input text content in LLM input messages.
     * Will mask information stored under the key `llm.input_messages.[i].message.contents.[j].message_content.text`.
     */
    private static final MaskingRule maskInputTextContentRule = new MaskingRule(
            args -> args.getConfig().isHideInputText()
                    && args.getKey().contains(SemanticConventions.LLM_INPUT_MESSAGES)
                    && args.getKey().contains(SemanticConventions.MESSAGE_CONTENT_TEXT),
            args -> REDACTED_VALUE);

    /**
     * Masks (redacts) output text content in LLM output messages.
     */
    private static final MaskingRule maskOutputTextContentRule = new MaskingRule(
            args -> args.getConfig().isHideOutputText()
                    && args.getKey().contains(SemanticConventions.LLM_OUTPUT_MESSAGES)
                    && args.getKey().contains(SemanticConventions.MESSAGE_CONTENT_TEXT),
            args -> REDACTED_VALUE);

    /**
     * Masks (removes) input images in LLM input messages.
     */
    private static final MaskingRule maskInputImagesRule = new MaskingRule(
            args -> args.getConfig().isHideInputImages()
                    && args.getKey().contains(SemanticConventions.LLM_INPUT_MESSAGES)
                    && args.getKey().contains(SemanticConventions.MESSAGE_CONTENT_IMAGE),
            args -> null);

    /**
     * Masks (redacts) base64 images that are too long.
     */
    private static final MaskingRule maskLongBase64ImageRule = new MaskingRule(
            args -> {
                Object value = args.getValue();
                if (value instanceof String) {
                    isBase64Url((String) value);
                }
                return false;
            },
            args -> REDACTED_VALUE);

    /**
     * Masks (removes) embedding vectors.
     */
    private static final MaskingRule maskEmbeddingVectorsRule = new MaskingRule(
            args -> args.getConfig().isHideInputEmbeddings()
                    && args.getKey().contains(SemanticConventions.EMBEDDING_EMBEDDINGS)
                    && args.getKey().contains(SemanticConventions.EMBEDDING_VECTOR),
            args -> null);

    /**
     * A list of masking rules that are applied to span attributes to either redact or remove sensitive information.
     * The order of these rules is important as it can ensure appropriate masking of information.
     * Rules should go from more specific to more general.
     */
    private static final List<MaskingRule> maskingRules = Arrays.asList(
            new MaskingRule(
                    args -> args.getConfig().isHideInputs() && args.getKey().equals(SemanticConventions.INPUT_VALUE),
                    args -> REDACTED_VALUE),
            new MaskingRule(
                    args -> args.getConfig().isHideInputs()
                            && args.getKey().equals(SemanticConventions.INPUT_MIME_TYPE),
                    args -> null),
            new MaskingRule(
                    args -> args.getConfig().isHideOutputs() && args.getKey().equals(SemanticConventions.OUTPUT_VALUE),
                    args -> REDACTED_VALUE),
            new MaskingRule(
                    args -> args.getConfig().isHideOutputs()
                            && args.getKey().equals(SemanticConventions.OUTPUT_MIME_TYPE),
                    args -> null),
            new MaskingRule(
                    args -> (args.getConfig().isHideInputs() || args.getConfig().isHideInputMessages())
                            && args.getKey().contains(SemanticConventions.LLM_INPUT_MESSAGES),
                    args -> null),
            new MaskingRule(
                    args -> (args.getConfig().isHideOutputs()
                                    || args.getConfig().isHideOutputMessages())
                            && args.getKey().contains(SemanticConventions.LLM_OUTPUT_MESSAGES),
                    args -> null),
            maskInputTextRule,
            maskOutputTextRule,
            maskInputTextContentRule,
            maskOutputTextContentRule,
            maskInputImagesRule,
            maskLongBase64ImageRule,
            maskEmbeddingVectorsRule);

    /**
     * Checks if a URL is a base64 encoded image
     */
    private static boolean isBase64Url(String url) {
        return url != null && url.startsWith("data:image/") && url.contains("base64");
    }

    /**
     * A function that masks (redacts or removes) sensitive information from span attributes based on the trace config.
     * @param args The masking rule arguments containing config, key, and value
     * @return The redacted value or null if the value should be masked, otherwise the original value
     */
    public static Value mask(MaskingRuleArgs args) {
        for (MaskingRule rule : maskingRules) {
            if (rule.getCondition().test(args)) {
                return rule.getAction().apply(args);
            }
        }
        return args.getValue();
    }

    /**
     * Convenience method for masking with individual parameters
     */
    public static Value mask(TraceConfig config, String key, Value value) {
        return mask(
                MaskingRuleArgs.builder().config(config).key(key).value(value).build());
    }
}
